### DISCUSSION TITLE
Coupling NEORV32 and AXI CDMA

### DISCUSSION THREAD

**Discussion by Sam-Vervaeck** [stnolting/neorv32#999] [Q&A]
I have coupled the NEORV32 to an AXI CDMA and it works mostly except for some "busy waiting" behaviour which I cannot figure out.

My test-case was to copy some data from one BRAM to another:
![image](https://github.com/user-attachments/assets/518f3ca7-05f7-43de-9246-79da7cd3cc05)

I have populated BRAM 0 with a COE file and after copying data to BRAM 1 I check if the same values are in the same relative memory offsets. For this functionality, I adapted the following function to set the relevant CDMA registers. Towards the end you can see a "busy waiting" section that checks the status register to see if the idle bit has already been set:

```
// This function initiates a DMA transaction. DMA copies num_bytes many bytes from src_addr to
// dest_addr. Both addresses are AXI addresses. 
// The function blocks until completion if block != 0. Otherwise it returns immediately.
void neorv32_cdma_transfer(uint32_t dest_addr, uint32_t src_addr, uint32_t num_bytes, uint8_t block)
{
	uint32_t block_val;

	//set source address:
	if(DEBUG) neorv32_uart0_printf("CDMA: setting source address to 0x%x\n", src_addr);
	CUSTOM_CDMA->CDMA_SA_reg = src_addr;
	if(DEBUG) neorv32_uart0_printf("                             is 0x%x\n", CUSTOM_CDMA->CDMA_SA_reg);

	//set dest address:
	if(DEBUG) neorv32_uart0_printf("CDMA: setting destin address to 0x%x\n", dest_addr);
	CUSTOM_CDMA->CDMA_DA_reg = dest_addr;
	if(DEBUG) neorv32_uart0_printf("                             is 0x%x\n", CUSTOM_CDMA->CDMA_DA_reg);

	//set number of transferred bytes. This also starts the transfer:
	CUSTOM_CDMA->CDMA_BTT_reg = num_bytes;
	if(DEBUG) neorv32_uart0_printf("CDMA: sending %u bytes\n", num_bytes);

	if(!block)
		return;

	//wait for completion:
	if(DEBUG) neorv32_uart0_printf("CDMA: waiting for completion of the transfer...\n");
	do
	{
		block_val = CUSTOM_CDMA->CDMASR_reg;
	} while(!(block_val & (1<<1)));

	if(DEBUG) neorv32_uart0_printf("CDMA: done waiting\n");
}
```

The whole point of a CDMA is that the NEO can continue working on other instructions while the CDMA does its thing. However, sometimes you need to wait on the data to have done copying so that is what the block parameter is for. When I created test code with the block parameter set to 1 and checked the waves, the address of the idle bit does not get checked during the CDMA execution? Weirdly enough, it only gets checked once after the CDMA has already been done for some time:

![image](https://github.com/user-attachments/assets/6dc120a4-533e-420e-af46-fb39eaf54194)

On the screenshot you can see the reading signals in blue and the writing in red. In the middle you can see the CDMA doing its copying and on the location of my yellow time marker the idle bit gets checked for the first time at register offset 0x04. I would have expected to see more reads on the AXI4-Lite status connection of the CDMA during execution according to my transaction C code?

Anybody has any experience with an AXI CDMA or a similar design and knows what is going on?

**Comment by stnolting** [comment]
What is the base address of the AXI CDMA? Have you enabled any NEORV32 caches?

**Reply by Sam-Vervaeck** [reply]
These are the hardware addresses:

![image](https://github.com/user-attachments/assets/075dabad-2027-442f-8c44-0e71b26eeabb)

These are the software addresses:

```
/** cdma module prototype */
typedef volatile struct __attribute__((packed, aligned(4))) {
    uint32_t CDMACR_reg;        // Offset 0x0, size 4 bytes
    uint32_t CDMASR_reg;        // Offset 0x4, size 4 bytes
    uint32_t reserved0[4];      // Padding to skip
    uint32_t CDMA_SA_reg;       // Offset 0x18, size 4 bytes
    uint32_t reserved1;         // Padding to skip
    uint32_t CDMA_DA_reg;       // Offset 0x20, size 4 bytes
    uint32_t reserved3;         // Padding to skip
    uint32_t CDMA_BTT_reg;      // Offset 0x28, size 4 bytes
} custom_cdma_t;

// Put at the same address as in the address editor in Vivado
#define CUSTOM_CDMA ((custom_cdma_t*) (0xC4000000))

/** BRAM 0 module prototype */
typedef volatile struct __attribute__((packed,aligned(4))) {
  uint32_t CONTROL; // offset 0, size 4 bytes
  uint32_t CONFIG;  // offset 4, size 4 bytes
} custom_BRAM_t;

// Put at the same address as in the address editor in Vivado
#define CUSTOM_BRAM_0 ((custom_BRAM_t*) (0xC0000000))

// Put at the same address as in the address editor in Vivado
#define CUSTOM_BRAM_1 ((custom_BRAM_t*) (0xC2000000))
```

No caching turned on at all and the only components are: `[NEORV32] Processor Configuration: IMEM DMEM XBUS MTIME UART0 CFS SYSINFO `

**Reply by stnolting** [reply]
Oh, okay. I thought some cache might be enabled so the accesses to AXI DMA got cached...

Your structs seem fine. They're marked as volatile, so you should see a bus transaction whenever you access one of the elements.

Does accessing any of the block RAMs from the core work without issues?

**Reply by Sam-Vervaeck** [reply]
It seems to work without issue.

I added this test code:
```
  neorv32_uart0_printf("Test access CUSTOM_BRAM_0->CONTROL: %x = 0x%x\n", (uint32_t) CUSTOM_BRAM_0, CUSTOM_BRAM_0->CONTROL);
  neorv32_uart0_printf("Test access CUSTOM_BRAM_1->CONTROL: %x = 0x%x\n", (uint32_t) CUSTOM_BRAM_1, CUSTOM_BRAM_1->CONTROL);
  neorv32_uart0_printf("Test access CUSTOM_CDMA->CDMASR_reg: %x = 0x%x\n", (uint32_t*) CUSTOM_CDMA + 1, CUSTOM_CDMA->CDMASR_reg);
```

And the prints and waves correspond:

```
Test access CUSTOM_BRAM_0->CONTROL: c0000000 = 0x36955295
Test access CUSTOM_BRAM_1->CONTRO: c2000000 = 0x00000000
Test access CUSTOM_CDMA->CDMASR_reg: c4000004 = 0x00000002
```

![image](https://github.com/user-attachments/assets/277c6aa6-86da-4f4f-9d28-d8128dade0f3)
![image](https://github.com/user-attachments/assets/b13f0ae5-874b-4792-a380-e10652545b19)
![image](https://github.com/user-attachments/assets/26617cd6-8f8d-4669-8c3e-088269ab893c)

**Reply by stnolting** [reply]
This looks fine. 🤔

One question regarding your `custom_BRAM_t`: the AXI block RAM controllers just make a transparent conversion from AXI to the native BRAM interface. But in case of your struct you are just using two block RAM entries (`CONTROL` and `CONFIG`) and not the entire 256k range that you have specified in your Vivado address map. Is this on purpose?

**Reply by Sam-Vervaeck** [reply]
Yeah that is just an artifact from a first setup in which I wanted easy access to memory locations for testing. I had created a memory mapped struct to make sure I did not have to work with number addresses at the beginning. Now, I have already expanded my testing code to work with raw addresses, e.g.:

```
  for (uint32_t i = 0; i<test_val; i+=800) {
    neorv32_uart0_printf("CUSTOM_BRAM_1 %x and %x = 0x%x%x\n", 0xC2000000 + (i*8+4), 0xC2000000 + (i*8), (uint32_t)(*((uint32_t *) (0xC2000000 + (i*8+4)))), (uint32_t)*((uint32_t *) (0xC2000000 + (i*8))));
   ...
```

**Reply by stnolting** [reply]
Oh ok ;)

Anyway, that should be irrelevant for the DMA issue.
Did you try a while-loop for polling the DMA state? Maybe something isn't right there (init value of `block_val`; or even the condition check itself? Is the `!` really correct?)...

```c
do
{
  block_val = CUSTOM_CDMA->CDMASR_reg;
} while(!(block_val & (1<<1)));
```

Rewritten as while-loop:

```c
while (CUSTOM_CDMA->CDMASR_reg & (1 << 1)); // keep polling until bit 1 clears
```

**Reply by Sam-Vervaeck** [reply]
The idle bit should actually not be cleared that's why the ! is there. If it is set to 1 the CDMA is idle, if it is 0 the CDMA is doing a transfer:
![image](https://github.com/user-attachments/assets/ccf2fbb3-da2d-4a41-bc82-7dc6f65e670d)

I have rewritten it as a while loop `	while (!(CUSTOM_CDMA->CDMASR_reg & (1<<1)));` and it gives the same waves

**Reply by stnolting** [reply]
> The idle bit should actually not be cleared that's why the ! is there. If it is set to 1 the CDMA is idle, if it is 0 the CDMA is doing a transfer:

Oh, you are right.

Have you checked your system in simulation or on the FPGA using an ILA?

**Reply by Sam-Vervaeck** [reply]
I have been working in simulation. I have a simple test bench that provides a clock to my block diagram wrapper.

**Reply by stnolting** [reply]
Hmm... 🤔

As far as I understand the CDMA's data sheet your code is just fine. I will try to test the CDMA on my side. Maybe someone else can help in the meantime?

**Reply by Sam-Vervaeck** [reply]
I will check with my thesis supervisors if they are willing to look at it for some time as well

**Reply by Sam-Vervaeck** [reply]
Hey! I have a solution update. I got the busy waiting doing its thing after playing around with (mainly) the printf statements:

(the blue wave part at the top is the AXI4-Lite)
![image](https://github.com/user-attachments/assets/57aff7b0-2896-4049-bd75-a45c2179cc41)

When turning these two printf statements off in between starting the CDMA and starting the busy waiting, it appears clearly on the previous waves screenshot:

```
        AXI_CDMA->CDMA_BTT_reg = num_bytes;
	// if(DEBUG) neorv32_uart0_printf("CDMA: sending %u bytes\n", num_bytes);

	if(!block)
		return;

	//wait for completion:
	// if(DEBUG) neorv32_uart0_printf("CDMA: waiting for completion of the transfer...\n");
	while (!(AXI_CDMA->CDMASR_reg & (1<<1)));
```

I created a larger CDMA transfer to time the duration of these prints and they seem to take 260 us:

![image](https://github.com/user-attachments/assets/7767b846-11ed-49db-ad7c-42f5ff3d1fe6)

So quite a bit longer than I had expected because that is 26000 clock cycles at 100 MHz if I am not mistaken. I suppose that has to do with the simulation of the UART (even though I am using the virtual sim one)?

**Reply by stnolting** [reply]
> I suppose that has to do with the simulation of the UART (even though I am using the virtual sim one)?

Well, `printf` takes some time as all the formatting and especially the number conversion is quite compute intense.

However, in your code the printing is not part of the DMA polling loop:
```c
while (!(AXI_CDMA->CDMASR_reg & (1<<1)));
```

So why should these printfs matter here? 🤔

**Reply by Sam-Vervaeck** [reply]
> So why should these printfs matter here? 🤔

So the reason why on my first screenshot all the way at the beginning of this thread there was no "active polling" visible on the rvalid and rready wires is because it polls for the first time when the CDMA is already idle for a while. This delay was because of the two printf statements which made the polling start after the data transfer was already completed. The first poll would then immediately see an idle CDMA and the while loop would be stopped.

To see some actual active polling, I had to make sure it starts while the data transfer is still busy by taking out those two printf statements.

**Reply by stnolting** [reply]
Ahh ok, now I understand.
And yes, that makes sense.

Anyway, good to hear that the system works as expected now. 🚀
